include/net/download_client.h

```diff
@@ -117,118 +152,72 @@ struct download_client_cfg {
-	bool set_tls_hostname;
+	size_t range_override;
+	/** Set socket to native TLS */
+	bool set_native_tls;
+	/** Keep connection to server when done */
+	bool keep_connection;
 };
 
 /**
- * @brief Download client asynchronous event handler.
- *
- * Through this callback, the application receives events, such as
- * download of a fragment, download completion, or errors.
- *
- * If the callback returns a non-zero value, the download stops.
- * To resume the download, use @ref download_client_start().
- *
- * @param[in] event	The event.
- *
- * @return Zero to continue the download, non-zero otherwise.
+ * @brief Download client state.
  */
-typedef int (*download_client_callback_t)(
-	const struct download_client_evt *event);
+enum download_client_state {
+	DOWNLOAD_CLIENT_DEINITIALIZED,
+	DOWNLOAD_CLIENT_IDLE,
+	DOWNLOAD_CLIENT_CONNECTING,
+	DOWNLOAD_CLIENT_CONNECTED,
+	DOWNLOAD_CLIENT_DOWNLOADING,
+	DOWNLOAD_CLIENT_DEINITIALIZING,
+};
 
 /**
  * @brief Download client instance.
+ *
+ * Members are set internally by the download client.
  */
 struct download_client {
-	/** Protect shared variables. */
-	struct k_mutex mutex;
-
-	/** Socket descriptor. */
-	int fd;
-
-	/** Destination address storage */
-	struct sockaddr remote_addr;
-
-	/** Response buffer. */
-	char buf[CONFIG_DOWNLOAD_CLIENT_BUF_SIZE];
-	/** Buffer offset. */
-	size_t offset;
-
+	/** Client configuration options. */
+	struct download_client_cfg config;
+	/** Host configuration options. */
+	struct download_client_host_cfg host_config;
+	/** Host name, null-terminated.
+	 */
+	char hostname[CONFIG_DOWNLOAD_CLIENT_MAX_HOSTNAME_SIZE];
+	/** File name, null-terminated.
+	 */
+	char file[CONFIG_DOWNLOAD_CLIENT_MAX_FILENAME_SIZE];
 	/** Size of the file being downloaded, in bytes. */
 	size_t file_size;
 	/** Download progress, number of bytes downloaded. */
 	size_t progress;
+	/** Buffer offset. */
+	size_t buf_offset;
 
-	/** Server hosting the file, null-terminated.
-	 *  The host name must be kept in scope while download is going on.
-	 */
-	const char *host;
-	/** File name, null-terminated.
-	 *  The file name must be kept in scope while download is going on.
+	/** Download client transport, http, CoAP, MQTT, ...
+	 *  Store a pointer to the selected transport per DLC instance to avoid looking it up each
+	 *  call.
 	 */
-	const char *file;
-	/** Configuration options. */
-	struct download_client_cfg config;
-
-	/** Protocol for current download. */
-	int proto;
-
-	struct  {
-		/** Whether the HTTP header for
-		 * the current fragment has been processed.
-		 */
-		bool has_header;
-		/** The server has closed the connection. */
-		bool connection_close;
-		/** Is using ranged query. */
-		bool ranged;
-	} http;
-
-	struct {
-		/** CoAP block context. */
-		struct coap_block_context block_ctx;
+	void *transport;
+	/** Transport parameters. */
+	uint8_t transport_internal[CONFIG_DOWNLOAD_CLIENT_TRANSPORT_PARAMS_SIZE];
```

Comment:
```suggestion
	struct transport_params transport_internal;
```

Also, I don't see the point of hiding those transports. You could have one struct where all possible types are in union.

```c
/** @cond INTERNAL_HIDDEN */
struct transport_params {
    union {
#if IS_ENABLED(CONFIG_DOWNLOAD_CLIENT_TRANSPORT_HTTP)
        struct transport_params_http http;
#endif
#if IS_ENABLED(CONFIG_DOWNLOAD_CLIENT_TRANSPORT_COAP)
        struct transport_params_coap coap;
#endif
    }
}
/** @endcond */
```

Now the transport parameter has exactly right size as compiler calculates it.

Reply:
The reason for doing it like this is that the application could add their own transport if desired, e.g. MQTT without changing the header.